File: 동적계획법/회장뽑기.py
"""
import sys
# sys.stdin = open("input.txt", 'r')

if __name__ == "__main__":
    n = int(input())
    dis = [[5000] * (n + 1) for _ in range(n + 1)]
    
    for i in range(1, n + 1):
        dis[i][i] = 0

    while True:
        i, j = map(int, input().split())
        if i == -1 and j == -1:
            break

        dis[i][j] = dis[j][i] = 1

    for k in range(1, n + 1):
        for i in range(1, n + 1):
            for j in range(1, n + 1):
                dis[i][j] = min(dis[i][j], dis[i][k] + dis[k][j])

    minimum = 2147000000
    for i in range(1, n + 1):
        max = 0
        for j in range(1, n + 1):
            if dis[i][j] > max:
                max = dis[i][j]
        if max < minimum:
            minimum = max

    cnt = 0
    candidate = []
    for i in range(1, n + 1):
        max = 0
        for j in range(1, n + 1):
            if dis[i][j] > max:
                max = dis[i][j]
        if minimum == max:
            cnt += 1
            candidate.append(i)

    print(minimum, cnt)
    print(' '.join(map(str, candidate)))
"""

import sys
# sys.stdin = open("input.txt", 'r')

if __name__ == "__main__":
    n = int(input())
    dis = [[1000] * (n + 1) for _ in range(n + 1)]

    for i in range(1, n + 1):
        dis[i][i] = 0

    while True:
        i, j = map(int, input().split())
        if i == -1 and j == -1: break

        dis[i][j] = dis[j][i] = 1

    for k in range(1, n + 1):
        for i in range(1, n + 1):
            for j in range(1, n + 1):
                dis[i][j] = min(dis[i][j], dis[i][k] + dis[k][j])

    result = [0] * (n + 1)
    score = 2147000000
    for i in range(1, n + 1):
        for j in range(1, n + 1):
            result[i] = max(result[i], dis[i][j])
        score = min(score, result[i])
    
    out = []
    for i in range(len(result)):
        if result[i] == score:
            out.append(i)

    print(score, len(out))
    print(' '.join(map(str, out)))

File: 동적계획법/회장뽑기.cpp
#include <bits/stdc++.h>
using namespace std;

int main(void){
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	// freopen("input.txt", "r", stdin);
	
	int n;
	int i, j, k, v1, v2;
	int maxi, mini, cnt;
	cin >> n;
	
	vector<vector<int> > graph(n + 1, vector<int> (n + 1, 1000));
	
	for(i = 1; i <= n; i++){
		graph[i][i] = 0;	
	}
	
	while(true){
		cin >> v1 >> v2;
		if(v1 == -1 && v2 == - 1) break;
		
		graph[v1][v2] = 1;
		graph[v2][v1] = 1;	
	}
	
	for(k = 1; k <= n; k++){
		for(i = 1; i <= n; i++){
			for(j = 1; j <= n; j++){
				graph[i][j] = min(graph[i][j], graph[i][k] + graph[k][j]);	
			}
		}	
	}
	
	vector<int> result(n + 1);
	mini = 2147000000;
	for(i = 1; i <= n; i++){
		maxi = 0;
		for(j = 1; j <= n; j++){
			if(graph[i][j] > maxi) maxi = graph[i][j];
		}
		result[i] = maxi;
		if(maxi < mini) mini = maxi;
	}
	cout << mini << " ";
	
	cnt = 0;
	for(i = 1; i <= n; i++){
		if(result[i] == mini){
			cnt++;	
		}
	}
	cout << cnt << "\n";
	for(i = 1; i <= n; i++){
		if(result[i] == mini) cout << i << " ";	
	}
	
	return 0;
}
